### PR TITLE
[Merged by Bors] - Avoid processing same ATX in parallel

### PR DESCRIPTION
**Pull request by poszu**
## Motivation
There is no mechanism preventing duplicated processing of ATX in parallel. An ATX might be gossiped multiple times while the same one is already being processed. Related issue: https://github.com/spacemeshos/go-spacemesh/issues/4426

## Changes
The ATX handler now registers the ATXID when it starts processing it. The subsequent calls to `handleATX()` will find that the handler is already processing this ATX and will register themself for the result, which is sent over a channel after the ATX processing is finished.

## Test Plan
Added a UT in which an ATX is gossiped 10x in parallel. The expectation is that it's processed only once.

### CODE REVIEW

File: activation/handler.go [hunk 1]
```diff
@@ -497,17 +504,47 @@ func (h *Handler) HandleGossipAtx(ctx context.Context, peer p2p.Peer, msg []byte
...
 
 func (h *Handler) handleAtx(ctx context.Context, expHash types.Hash32, peer p2p.Peer, msg []byte) error {
-	receivedTime := time.Now()
 	var atx types.ActivationTx
 	if err := codec.Decode(msg, &atx); err != nil {
 		return fmt.Errorf("%w: %w", errMalformedData, err)
 	}
-
-	atx.SetReceived(receivedTime.Local())
+	atx.SetReceived(time.Now().Local())
```
Comment (fasmat) [inline activation/handler.go]: why the change? this shifts the time it was received to the time it was received + decoded?

File: activation/handler.go [hunk 2]
```diff
@@ -497,17 +504,47 @@ func (h *Handler) HandleGossipAtx(ctx context.Context, peer p2p.Peer, msg []byte
...
+	h.inProgressMu.Lock()
+	if sub, ok := h.inProgress[atx.ID()]; ok {
+		ch := make(chan error, 1)
+		h.inProgress[atx.ID()] = append(sub, ch)
+		h.inProgressMu.Unlock()
+		h.log.WithContext(ctx).With().Debug("atx is already being processed. waiting for result", atx.ID())
+		select {
+		case err := <-ch:
+			h.log.WithContext(ctx).With().Debug("atx processed in other task", atx.ID(), log.Err(err))
+			return err
```
Comment (fasmat) [inline activation/handler.go:527]: I'm not sure if returning the same `error` for all gossiped copies of the same ATX is correct? Could it be that an ATX fails validation now but might pass later (e.g. missing dependencies are now available)? 
Also if `error` is not `nil` that would be suspicious - that would mean multiple nodes gossiped the ATX and consider it valid.

Would it make sense not to return `err` here but instead continue validation as normal? If the ATX was processed already it will exit early with `errKnown` otherwise it will try to validate again.

**Review by fasmat — APPROVED** [review]

File: activation/handler.go (hunk 2 above)
Comment (poszu) [reply activation/handler.go:527]: > Could it be that an ATX fails validation now but might pass later (e.g. missing dependencies are now available)?

I'm not sure about that. The peer that gossiped the invalid ATX should have all its dependencies already. Some transient networking issues could cause it.

I don't think it makes sense to disregard the error and retry validation. There is little chance it will succeed the second time. That ATX should be re-gossiped (if it comes from "the owner") or we will fetch it when needed.

**Comment by poszu** [comment]
bors try

File: activation/handler.go (hunk 2 above)
Comment (fasmat) [reply activation/handler.go:527]: > The peer that gossiped the invalid ATX should have all its dependencies already.

Yes but the peer that receives the ATX and validates it might not and then fail the validation, when it could pass it at a later point (after receiving the dependencies for it).

Comment (poszu) [reply activation/handler.go:527]: The node that processes a gossiped ATX proactively fetches its dependencies here: https://github.com/spacemeshos/go-spacemesh/blob/20194e8e76441a12376b5dda36abed9d381bff42/activation/handler.go#L563

It means that the peer that gossiped the ATX doesn't have the deps if that fails, which means something about that node is utterly wrong.

Comment (poszu) [reply activation/handler.go:527]: Actually, there is no guarantee that `FetchReferences()` would ask the same peer that gossiped the ATX. It will ask a random peer, so a fetch request could fail if a given peer doesn't have the dep yet. This is suboptimal and could be improved in the future. 

Anyway, the _fetcher_ has a retry mechanism built-in - it will retry up to 100 times by default. It seems unlikely for `FetchReferences()` to fail if the dep exists somewhere to be fetched. Wdyt?

**Comment by poszu** [comment]
bors merge
